simpl/dicts.py

```diff
@@ -0,0 +1,133 @@
+# Copyright (c) 2011-2015 Rackspace US, Inc.
+# All Rights Reserved.
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+"""Helpers for built-in `dict` class."""
+
+
+def write_path(target, path, value, separator='/'):
+    """Write a value deep into a dict building any intermediate keys.
+
+    :param target: a dict to write data to
+    :param path: a key or path to a key (path is delimited by `separator`)
+    :param value: the value to write to the key
+    :keyword separator: the separator used in the path (ex. Could be "." for a
+        json/mongodb type of value)
+    """
+    parts = path.split(separator)
+    current = target
+    for part in parts[:-1]:
+        if part not in current:
+            current[part] = current = {}
+        else:
+            current = current[part]
+    current[parts[-1]] = value
+
+
+def read_path(source, path, separator='/'):
+    """Read a value from a dict supporting a deep path as a key.
+
+    :param source: a dict to read data from
+    :param path: a key or path to a key (path is delimited by `separator`)
+    :keyword separator: the separator used in the path (ex. Could be "." for a
+        json/mongodb type of value)
+    """
+    parts = path.strip(separator).split(separator)
+    current = source
+    for part in parts[:-1]:
+        if part not in current:
+            return
+        current = current[part]
+        if not isinstance(current, dict):
+            return
+    return current.get(parts[-1])
+
+
+def path_exists(source, path, separator='/'):
+    """Check a dict for the existence of a value given a path to it.
+
+    :param source: a dict to read data from
+    :param path: a key or path to a key (path is delimited by `separator`)
+    :keyword separator: the separator used in the path (ex. Could be "." for a
+        json/mongodb type of value)
+    """
+    if path == separator and isinstance(source, dict):
+        return True
+    parts = path.strip(separator).split(separator)
+    if not parts:
+        return False
+    current = source
+    for part in parts:
+        if not isinstance(current, dict):
+            return False
+        if part not in current:
+            return False
+        current = current[part]
+    return True
+
+
+def merge_dictionary(dst, src, extend_lists=False):
```

Comment:
The semantics of `merge_dictionary` are really strange; I'm not even sure how I could use this. Given the examples below, if I had to "merge" two dictionaries, I can't comprehend what I would expect to happen.

```
>>> from simpl import dicts
>>> d1 = dict(a=[1,2]); d2=dict(a=[2,1]); dicts.merge_dictionary(d1, d2, extend_lists=False)
{'a': [2, 1]}
>>> d1 = dict(a=[1,2]); d2=dict(a=[2,1]); dicts.merge_dictionary(d1, d2, extend_lists=True)
{'a': [1, 2]}
>>> d1 = dict(a=[1,2]); d2=dict(a=[2,1]); dicts.merge_dictionary(d2, d1, extend_lists=False)
{'a': [1, 2]}
>>> d1 = dict(a=[1,2]); d2=dict(a=[2,1]); dicts.merge_dictionary(d2, d1, extend_lists=True)
{'a': [2, 1]}
>>> d1 = dict(a=[1,None,2]); d2=dict(a=[None,2,1]); dicts.merge_dictionary(d1, d2, extend_lists=False)
{'a': [1, 2, 1]}
>>> d1 = dict(a=[1,None,2]); d2=dict(a=[None,2,1]); dicts.merge_dictionary(d1, d2, extend_lists=True)
{'a': [1, None, 2]}
>>> d1 = dict(a=[1,None,2]); d2=dict(a=[None,2,1]); dicts.merge_dictionary(d2, d1, extend_lists=False)
{'a': [1, 2, 2]}
>>> d1 = dict(a=[1,None,2]); d2=dict(a=[None,2,1]); dicts.merge_dictionary(d2, d1, extend_lists=True)
{'a': [None, 2, 1]}
>>> 
```

Can you explain this behavior? It almost seems like the lists are behaving like some kind ordered sets--sometimes. Sometimes the `dst` dominates the order and contents of the list, sometimes `None` values are dropped.

Does `src` override `dst` if a key in a nested dict exists in both, but with a different value?

```
>>> d1 = dict(a=dict(foo='baz')); d2=dict(a=dict(foo='bar')); dicts.merge_dictionary(d2, d1, extend_lists=False)
{'a': {'foo': 'baz'}}
>>> d1 = dict(a=dict(foo='baz')); d2=dict(a=dict(foo='bar')); dicts.merge_dictionary(d1, d2, extend_lists=False)
{'a': {'foo': 'bar'}}
>>> 
```